Fix/roles

afaikafk opened the pull request:
This PR actually contains UX improvements, and restricts the view of status queue to moderators only.

afaikafk committed:
Assign real role values

afaikafk committed:
Hide show full status link for non-moderators

afaikafk committed:
Add refresh-interval option for roca widgets

afaikafk committed:
Set up roca autorefresh on main and status page of current bin status

afaikafk committed:
Add xhr method whitelist to RoleMixin class

afaikafk committed:
Apply xhr whitelist to Status endpoint such that partial status requests can be allowed while everything else restricted to moderators

afaikafk committed:
Expand roca refresh methods to refresh container when specific events are fired, optinally delayed with a specified amount of time

afaikafk committed:
Refresh upload widget and both file lists when a file is uploaded, with file widget refresh being delayed

afaikafk committed:
Make function a one-liner

foxbunny commented on broadcast/util/routes.py:
```diff
@@ -381,10 +385,17 @@ def get_role_needs_login_message(self):
         if self.request.method in self.get_role_method_whitelist():
             return
 
+        if (self.request.is_xhr and
```
EDIT: Ok, scratch this. I should have some coffee. :P

foxbunny commented on broadcast/skins/broadcast/src/coffee/roca.coffee:
```diff
@@ -35,16 +35,21 @@
 
+    submitComplete = () ->
+      if options && options.onCompleteEvent
+        win.trigger options.onCompleteEvent
```
Isn't it much easier to just use the element ID as (part of) the event name?

foxbunny commented on broadcast/skins/broadcast/src/coffee/roca.coffee:
```diff
@@ -95,7 +113,17 @@
       if (el.data 'roca-trap-submit') is 'yes'
-        target.funnelSubmit()
+        options = { "onCompleteEvent": el.data 'roca-submit-complete-event' }
+        target.funnelSubmit(options)
```
So, instead of hard-wiring refresh to `funnelSubmit`, make it a generic event handler. Skip passing complete event name here, and see comment further below.

foxbunny commented on broadcast/skins/broadcast/src/coffee/roca.coffee:
```diff
@@ -95,7 +113,17 @@
+      event = el.data 'roca-refresh-on'
+      if event
+        delay = el.data 'roca-refresh-delay'
+        refreshOn target, event, delay
```
This is all that is needed for the auto refresh. What event is triggered and how, is not this plugin's concern. This plugin should just worry about what even to listen for.

foxbunny commented on broadcast/skins/broadcast/src/coffee/roca.coffee:
```diff
@@ -95,7 +113,17 @@
+      # a specified time
+      event = el.data 'roca-refresh-on'
+      if event
+        delay = el.data 'roca-refresh-delay'
```
If we are doing something on a timer, then it's probably wrong. I suggest dropping the delay altogether, and _manually_ triggering the event in the form handler as need.

foxbunny commented on broadcast/skins/broadcast/src/coffee/roca.coffee:
```diff
@@ -99,8 +99,7 @@
         _delay = (delay || 0) * 1000
-        _refresh = () ->
-           target.loading().reload()
+        _refresh = () -> target.loading().reload()
```
Do _not_ use loading() when calling reload().

foxbunny commented:
As a general comment, one thing I've learned during implementation of post-vote scroll is that synchronizing events using a timer is not worth considering as an option. As a general rule, let's not use timers for syncing events.

It is much safer to sync on concrete events, like XHR completion even if it means the code grows from 3 lines to a dozen. It provides a much smoother UX and is not subject to weird race conditions and edge cases (such as request taking longer than expected). The only place where timeout syncing is inevitable is syncing with CSS animation, as it's not something JavaScript can tap into.

Also, my gut feeling is that auto-refresh is already bordering on bloat in the ROCA. Refresh on event is a generic feature that is quite a good fit for the plugin, but auto-refresh is already in application-specific domain, and should be handled by an external driver firing events.

foxbunny commented:
Oh, and auto-refresh doesn't need to be on implemented in a loop. The last update thing I was talking about can be used for this, and stats refreshed only when there's a change in the last update timestsamp.

integricho commented:
Could you apply the changes that you mentioned here?

foxbunny commented:
I'm already doing some of it. Not sure how much of it I'll be able to pull off by Monday, though.

Commit:
Use XHRJsonRoute for LastUpdate and return dict

This commit changes the LastUpdate route such that the entire last_update dict
is returned as JSON instead of just the timestamp as string.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Make last_update timestamp available to templates

This commit makes the last_update dict available to templates, and rigs the
base template to expose the timestamp value to JavaScript code via a
data-property.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Add last update timestamp synchronizer

This commit adds a new JS module, last-update, which syncrhonizes the last
update timestamp between the server and client, and fires a 'state-update'
event globally when the timestamp changes.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Form submit events fix

This commit changes the form submission handler such that event name is not
acquired via options, but rather chosen based on the container's id. Also
changes the submit event trigger such that it's triggered on successful
submission only.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Trigger ROCA container-specific load/error events

This patch adds container-specific load/error events in <id>-roca-load and
<id>-roca-error format. This allows listeners to choose specific containers to
listen to.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Pass container element along with submit event

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Load guard for rocaLoad() method

This commit adds a simple check that prevents multiple calls to rocaLoad from
creating a race condition.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Remove autorefresh and submit event passing

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Fix typo

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Pass status data when triggering the event

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Fac out container setup, move attrs to container

This commit moves all attributes except roca-target to the actual container.
This, in turn, allows us to set up roca containers even in situations where
there is no anchor.

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Redo ROCA attribs to match the new ROCA API

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Add success attrib to feedback template

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Present a pop up when queue state changes

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Reload upload form on submit

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Recompile main JS module

Signed-off-by: Branko Vukelic <branko@outernet.is>

Commit:
Update timstamp when new item is created

Signed-off-by: Branko Vukelic <branko@outernet.is>
